api/src/main/java/org/openmrs/api/context/Context.java

```diff
@@ -177,6 +182,32 @@ public static void setDAO(ContextDAO dao) {
 		contextDAO = dao;
 	}
 
+	/**
+	 * Spring init method that sets the authentication scheme.
+	 */
+	static private void setAuthenticationScheme() {
+
+		authenticationScheme = new UsernamePasswordAuthenticationScheme();
+
+		try {
+			authenticationScheme = Context.getServiceContext().getApplicationContext().getBean(AuthenticationScheme.class); // manual autowiring (from a module)
+			log.info("An authentication scheme override was provided. Using this one in place of the OpenMRS default authentication scheme.");
+		}
+		catch(NoUniqueBeanDefinitionException e) {
+			log.error("Multiple authentication schemes overrides are being provided, this is currently not supported. Sticking to OpenMRS default authentication scheme.");
+		}
+		catch(NoSuchBeanDefinitionException e) {
+			log.debug("No authentication scheme override was provided. Sticking to OpenMRS default authentication scheme.");
+		}
+		catch(BeansException e){
+			log.error("Fatal error encountered when injecting the authentication scheme override. Sticking to OpenMRS default authentication scheme.");
+		}
+
+		if (authenticationScheme instanceof DaoAuthenticationScheme) {
```

Comment:
A special case of setting the contextDAO from the setAuthenticationScheme() method when the scheme is a DaoAuthenticationScheme does not look very intuitive to me. 😊 
Can't the contextDAO be dependency injected or set from somewhere else?

Reply:
Should make all schemes subclass `DaoAuthenticationScheme`, thereby perhaps renaming it `BaseAuthenticationScheme`?

Until here I sort of left the question open as to whether some schemes might not need `ContextDAO`, but maybe this will just never happen in OpenMRS. I guess this might one day be bypassed if OpenMRS was authenticating using Spring Security instead of handling all those things "manually". What do you think?

Reply:
Why would other schemes need ContextDAO?

Reply:
While the actual authentication process will be delegated to a third-party authority (an OAuth2 provider such as Keycloak typically), OpenMRS still requires one of its own users. Think of the roles for example, it is not possible to just pass a `User` instance here and there saying "it's authenticated", OpenMRS will connect roles to it, and for this it is necessary to have a `User` entity to exist in OpenMRS' database as well.

The question is rather... will there be a point when it will be possible to not rely on `ContextDAO` anymore when delegating authentication elsewhere?

Reply:
Does this mean the roles will also need to be inserted into the OpenMRS database?

Reply:
No, we assume that the roles we need already exist. So we might add roles to the new user, but we won't create those roles.

Reply:
Sorry, that is exactly what i did mean. In other words, i should have asked like this: "Are you also inserting user to role mappings?"